src/providers/tpm_provider/mod.rs

```diff
@@ -0,0 +1,462 @@
+        {
+            error!("The TPM provider currently only supports creating RSA key pairs for signing and verifying.");
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+
+        let key_name = op.key_name;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+        // This should never panic on 32 bits or more machines.
+        let key_size = std::convert::TryFrom::try_from(op.key_attributes.key_size)
+            .expect("Conversion to usize failed.");
+
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let (key_context, auth_value) = esapi_context
+            .create_rsa_signing_key(key_size, AUTH_VAL_LEN)
+            .or_else(|e| {
+                error!("Error creating a RSA signing key: {}.", e);
+                dbg!(e);
+                Err(ResponseStatus::PsaErrorHardwareFailure)
+            })?;
+
+        insert_password_context(
+            &mut *store_handle,
+            key_triple,
+            PasswordContext {
+                context: key_context,
+                auth_value,
+            },
+        )?;
+
+        Ok(ResultCreateKey {})
+    }
+
+    fn import_key(&self, app_name: ApplicationName, op: OpImportKey) -> Result<ResultImportKey> {
+        if op.key_attributes.key_type != KeyType::RsaPublicKey
+            || op.key_attributes.algorithm != Algorithm::sign(SignAlgorithm::RsaPkcs1v15Sign, None)
+        {
+            error!(
+                "The TPM provider currently only supports importing RSA public key for verifying."
+            );
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+
+        let key_name = op.key_name;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+        let key_data = op.key_data;
+
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let public_key: RsaPublicKey = serde_asn1_der::from_bytes(&key_data).or_else(|err| {
+            error!("Could not deserialise key elements: {}.", err);
+            Err(ResponseStatus::PsaErrorCommunicationFailure)
+        })?;
+        if public_key.public_exponent.to_bytes_be().1 != PUBLIC_EXPONENT {
+            error!("The TPM Provider only supports 0x101 as public exponent for RSA public keys, {:?} given.", public_key.public_exponent.to_bytes_be());
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+        let key_data = public_key.modulus.to_bytes_be().1;
+
+        let len = key_data.len();
+        if len != 128 && len != 256 {
```

Comment:
I might've mislead you, not sure that's actually true, will look into it

Reply:
I think the TPM can support whatever sizes it wants (larger than 1024)

Reply:
I will only keep the lower bound check then!